sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -178,6 +179,142 @@ case class Like(left: Expression, right: Expression, escapeChar: Char)
   }
 }
 
+abstract class LikeAllBase extends Expression with ImplicitCastInputTypes {
```

Comment:
Could we make it only support `Literal`, for example:
```scala
case class LikeAll(child: Expression, isNotDefined: Boolean, seq: mutable.Buffer[Any])
  extends UnaryExpression with ImplicitCastInputTypes with NullIntolerant {

  override def dataType: DataType = BooleanType

  override def inputTypes: Seq[DataType] = StringType :: Nil

  @transient private[this] lazy val hasNull: Boolean = seq.contains(null)

  @transient private lazy val cachedPattern = seq.filterNot(_ == null)
    .map(s => Pattern.compile(StringUtils.escapeLikeRegex(s.toString, '\\')))

  override protected def nullSafeEval(input1: Any): Any = {
    if (hasNull) {
      false
    } else {
      val str = input1.asInstanceOf[UTF8String].toString
      if (isNotDefined) {
        !cachedPattern.exists(p => p.matcher(str).matches())
      } else {
        cachedPattern.forall(p => p.matcher(str).matches())
      }
    }
  }

 // TODO: codegen
}
```

```scala
val exps = ctx.expression.asScala.map(expression)
validate(exps.nonEmpty, "Expected something between '(' and ')'.", ctx)
if (exps.size > 10 && exps.forall(_.foldable)) {
   LikeAll(e, isNotDefined, exps.map(_.eval(EmptyRow)))
} else {
  exps.map(p => invertIfNotDefined(Like(e, p))).reduceLeft(And)
}
```

Reply:
Sounds reasonable. If there are more than 14378 elements, most likely they are literals.

Reply:
The current implementation requires the expression list to be foldable, including literal. In addition, in my earliest implementation, `nullSafeEval` also used the caching of each pattern.  But through offline discussions with @cloud-fan , there is no need to do this. The current implementation of `doGenCode`, if it is all literal, has actually achieved the effect.